rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2973 @@
+
+#### Common Use-Case: As-Stable-As-Possible
+[stability-common]: #common-use-case-as-stable-as-possible
+To promise that all transmutations which are currently safe for your type will remain so in the future, simply annotate your type with:
+```rust
+#[derive(PromiseTransmutableFrom, PromiseTransmutableInto)]
+#[repr(C)]
+pub struct Foo(pub Bar, pub Baz);
+```
+This expands to:
+```rust
+#[repr(C)]
+pub struct Foo(pub Bar, pub Baz);
+
+/// Generated `PromiseTransmutableFrom` for `Foo`
+const _: () = {
+    use core::transmute::stability::PromiseTransmutableFrom;
+
+    #[repr(C)]
+    pub struct TransmutableFromArchetype(
+        pub <Bar as PromiseTransmutableFrom>::Archetype,
+        pub <Baz as PromiseTransmutableFrom>::Archetype,
+    );
+
+    impl PromiseTransmutableFrom for Foo {
+        type Archetype = TransmutableFromArchetype;
+    }
+};
+
+/// Generated `PromiseTransmutableInto` for `Foo`
+const _: () = {
+    use core::transmute::stability::PromiseTransmutableInto;
+
+    #[repr(C)]
+    pub struct TransmutableIntoArchetype(
+        pub <Bar as PromiseTransmutableInto>::Archetype,
+        pub <Baz as PromiseTransmutableInto>::Archetype,
+    );
+
+    impl PromiseTransmutableInto for Foo {
+        type Archetype = TransmutableIntoArchetype;
+    }
+};
+```
+Since deriving *both* of these traits together is, by far, the most common use-case, we [propose][extension-promisetransmutable-shorthand] `#[derive(PromiseTransmutable)]` as an ergonomic shortcut.
+
+#### Uncommon Use-Case: Weak Stability Guarantees
+[stability-uncommon]: #uncommon-use-case-weak-stability-guarantees
+
+If you are *most people* and want to declare your types as layout-stable, you should follow the advice in the previous sections. In doing so, you declare that you will *not* modify the layout of your type in virtually any way, except as a breaking change. If your type's fields have simple stability guarantees, this effects the strongest possible declaration of stability: it declares that *all* transmutations that are safe are *also* stable.
+
+However, if you anticipate that you will need to make changes to your type's layouts in a non-breaking way, you can craft uninformative archetypes that effect *weaker* promises of stability. With uninformative archetypes, only a *subset* of transmutations that are safe are also considered stable. For instance, you may use `PromiseTransmutableInto` to specify an lower-bound for the size of a type, but *not* make promises about its bit-validity:
+
+```rust
+#[repr(C)] pub struct Foo(u8, u8, u8, u8, u8, u8);
+
+#[repr(C)]
+struct FooArchetype(pub [MaybeUninit<u8>; 5]);
+
+impl PromiseTransmutableInto for Foo {
+    type Type = [MaybeUninit<u8>; 5];
+}
+```
+
+##### Crafting Uninformative `PromiseTransmutableFrom` Archetypes
```

Comment:
I loved reading this and I'm so excited about it. For me this section was the hardest part to wrap my head around, I kept thinking that some of the language in here might be the wrong way around for what it's discussing. I still don't grok it enough to actually know if any of it is the wrong way around without spending like 30 minutes drawing diagrams. Which means my actual comment is not necessarily that there's anything wrong here but that this section might benefit from a diagram or some other wording to make this easier to follow.

Reply:
I've [revamped this section](https://github.com/jswrenn/project-safe-transmute/blob/rfc/rfcs/0000-safe-transmute.md#uncommon-use-case-weak-stability-guarantees), replacing it with an example focusing in-depth on alignment!